### PR TITLE
fix: prevent division by zero in ClippedPGLossFn calculation

### PR DESCRIPTION
**Pull request by zpqiu**
# What does this PR do ?

Prevent division by zero in ClippedPGLossFn calculation when mask.sum() = 0.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):
#158 

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA/reinforcer/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/reinforcer/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/reinforcer/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
NA

### CODE REVIEW

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 1]
```diff
@@ -127,13 +129,14 @@ def __call__(
         if mask.sum() > 0:
             actor_loss = masked_mean(torch.max(loss1, loss2), mask)
             loss = actor_loss + kl
+            with torch.no_grad():
+                probs_ratio = masked_mean(ratios.detach(), mask).item()
```
Comment (SahilJain314) [inline nemo_reinforcer/algorithms/loss_functions.py]: Since you've guarded the masked_mean here, I assume that function fails with a mask sum of 0? If so, can you please guard that inside of masked_mean? maybe do a quick check inside of it if a 'check for zero mask' arg is true. 
Should return sum * 0 (not just 0) in that case. (maintain grad graph)

I see places in this file that it isn't guarded (i.e. L112) and it just seems like a really easy place to trip and fall.

Comment (SahilJain314) [reply nemo_reinforcer/algorithms/loss_functions.py]: Also, can you please add a unit test for the all-zeros case in `tests/unit/algorithms/test_loss_functions.py`

File: nemo_reinforcer/algorithms/utils.py
Comment (zpqiu) [inline nemo_reinforcer/algorithms/utils.py:1]: I'm unsure whether it's expected behavior for `as_masked_tensor(values, mask.bool()).mean(dim=dim).to_tensor(torch.nan)` to return nan when the mask tensor's dimension contains all-zero entries. 

Another implementation approach for both mean ops, similar to other frameworks, would look like this code:
```python
def masked_mean(values, mask, axis=None):
    """Compute mean of tensor with a masked values."""
    return (values * mask).sum(axis=axis) / (mask.sum(axis=axis) + 1e-8)
```

File: nemo_reinforcer/algorithms/loss_functions.py (hunk 1 above)
Comment (zpqiu) [reply nemo_reinforcer/algorithms/loss_functions.py]: I reimplemented masked_mean, but there are still some issues that I haven't figured out.

File: nemo_reinforcer/algorithms/utils.py
Comment (SahilJain314) [reply nemo_reinforcer/algorithms/utils.py:1]: I'm happy with this eps formulation to solve the div 0 problem. I'm not sure what you mean by the nan issue above though.

Comment (zpqiu) [reply nemo_reinforcer/algorithms/utils.py:1]: > I'm unsure whether it's expected behavior for as_masked_tensor(values, mask.bool()).mean(dim=dim).to_tensor(torch.nan) to return nan when the mask tensor's dimension contains all-zero entries.

Since I am not familiar with the `as_masked_tensor` function used in the original code, I don't know how to handle the case where mask tensor partially contains all zero entries, like [[1, 0, 0], [0, 0, 0], [0, 0, 1]].

> I'm happy with this eps formulation to solve the div 0 problem. I'm not sure what you mean by the nan issue above though.

I also think the second approach using eps is simple and effective. I will revise the masked_mean function accordingly.

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 2]
```diff
@@ -91,7 +91,7 @@ def __call__(
         mask = token_mask * sample_mask.unsqueeze(-1)
 
         lp_error = torch.abs(generation_logprobs - prev_logprobs)  # noqa: F841  (precommit ignore for now)
-        mult_prob_error = ((torch.exp(lp_error) * mask).sum() / mask.sum()).item()
+        mult_prob_error = masked_mean(lp_error, mask).item()
```
Comment (SahilJain314) [inline nemo_reinforcer/algorithms/loss_functions.py]: we need the 'exp' here. can you put that back?

Comment (zpqiu) [reply nemo_reinforcer/algorithms/loss_functions.py]: Sorry for this mistake. Have fixed.

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 3]
```diff
@@ -124,13 +124,8 @@ def __call__(
...
 
-        if mask.sum() > 0:
-            actor_loss = masked_mean(torch.max(loss1, loss2), mask)
-            loss = actor_loss + kl
-        else:
-            # disable this update since there are no valid tokens
-            loss = loss1.view(-1)[0] * 0
-
+        actor_loss = masked_mean(torch.max(loss1, loss2), mask)
+        loss = actor_loss + kl
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/loss_functions.py:128]: previously, the KL was skipped if no valid tokens. Here it looks like KL is still part of the loss if nothing valid. Is that intended?

Comment (zpqiu) [reply nemo_reinforcer/algorithms/loss_functions.py:128]: > previously, the KL was skipped if no valid tokens. Here it looks like KL is still part of the loss if nothing valid. Is that intended?

In my understanding, if there are no valid tokens, masked_mean(kl) will be 0.

Comment (terrykong) [reply nemo_reinforcer/algorithms/loss_functions.py:128]: ah yes! you're right. I missed it in the diff above this

File: nemo_reinforcer/algorithms/utils.py [hunk 4]
```diff
@@ -120,9 +120,7 @@ def wrapper(*args, **kwargs):
 @surpress_user_warnings
 def masked_mean(values, mask, dim=None):
     """Masks values with mask, and computes the mean of the values using the masked values."""
-    if dim is None:
-        return values[mask.bool()].mean()
-    return as_masked_tensor(values, mask.bool()).mean(dim=dim).to_tensor(torch.nan)
+    return (values * mask).sum(dim=dim) / (mask.sum(dim=dim) + 1e-8)
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/utils.py:123]: image

The GRPO loss from the paper loops over valid seqlen first and then over the generations (so seqlen then batchdim). Should we default to that for this function?

Comment (zpqiu) [reply nemo_reinforcer/algorithms/utils.py:123]: DAPO and Dr.GRPO have already explained that seq-mean-batch-mean may have some problems. 
And, as @SahilJain314 suggested, we can create a new PR to add a flag to switch between different mean approaches.
